File: src/ate_cloud/schemas/spc.py
"""Pydantic schemas for SPC (Statistical Process Control) resources.

Defines response models for:
- SPCStatistics: capability indices (Cp, Cpk, Ppk) and process metrics
- SPCChart: X-bar / R control chart data with control limits
- SPCAlert: anomaly alert (Western Electric rule violation or Ppk threshold)
"""

from datetime import datetime

from pydantic import BaseModel, Field


class SPCStatistics(BaseModel):
    """SPC capability indices and process statistics for one stream.

    Attributes:
        product_type: Product type identifier.
        measurement_name: Measurement name.
        sample_count: Number of samples in the sliding window.
        mean: Arithmetic mean of values.
        std_dev_within: Within-subgroup sigma (R-bar / d2), used for Cp/Cpk.
        std_dev_overall: Overall sigma (population), used for Ppk.
        cp: Process capability index (potential).
        cpk: Process capability index (actual, within-subgroup).
        ppk: Preliminary process performance index (overall sigma).
        usl: Upper specification limit.
        lsl: Lower specification limit.
        last_updated: Timestamp of the last sample processed.
    """

    product_type: str
    measurement_name: str
    sample_count: int = Field(ge=0)
    mean: float | None = None
    std_dev_within: float | None = None
    std_dev_overall: float | None = None
    cp: float | None = None
    cpk: float | None = None
    ppk: float | None = None
    usl: float | None = None
    lsl: float | None = None
    last_updated: datetime | None = None


class SPCSubgroupStat(BaseModel):
    """One subgroup's X-bar and R statistics.

    Attributes:
        index: Subgroup ordinal (0-based).
        mean: Subgroup mean (X-bar).
        range: Subgroup range (max - min).
        sample_count: Number of samples in the subgroup.
    """

    index: int
    mean: float
    range: float
    sample_count: int


class SPCChart(BaseModel):
    """X-bar / R control chart data with control limits.

    Attributes:
        product_type: Product type identifier.
        measurement_name: Measurement name.
        center_line: X-bar grand average (center line of X-bar chart).
        ucl: Upper control limit for X-bar chart.
        lcl: Lower control limit for X-bar chart.
        r_center: Center line for R chart (R-bar).
        r_ucl: Upper control limit for R chart.
        r_lcl: Lower control limit for R chart (0 for n<=6).
        subgroup_size: Number of samples per subgroup.
        subgroups: List of subgroup statistics.
    """

    product_type: str
    measurement_name: str
    center_line: float | None = None
    ucl: float | None = None
    lcl: float | None = None
    r_center: float | None = None
    r_ucl: float | None = None
    r_lcl: float | None = None
    subgroup_size: int = 5
    subgroups: list[SPCSubgroupStat] = Field(default_factory=list)


class SPCAlert(BaseModel):
    """An SPC alert raised by anomaly detection.

    Attributes:
        product_type: Product type identifier.
        measurement_name: Measurement name.
        rule: Rule that triggered (e.g. 'WE1_beyond_3sigma', 'Ppk_below_1.00').
        severity: 'warning' or 'critical'.
        message: Human-readable description.
        value: The value that triggered the alert (if applicable).
        timestamp: When the alert was generated.
        sample_count: Number of samples seen when the alert fired.
    """

    product_type: str
    measurement_name: str
    rule: str
    severity: str
    message: str
    value: float | None = None
    timestamp: datetime
    sample_count: int

File: frontend/src/api/spc.ts
import http from './interceptor'

const api = http

/**
 * SPC capability indices and process statistics for one stream.
 */
export interface SPCStatistics {
  product_type: string
  measurement_name: string
  sample_count: number
  mean: number | null
  std_dev_within: number | null
  std_dev_overall: number | null
  cp: number | null
  cpk: number | null
  ppk: number | null
  usl: number | null
  lsl: number | null
  last_updated: string | null
}

/**
 * One subgroup's X-bar and R statistics.
 */
export interface SPCSubgroupStat {
  index: number
  mean: number
  range: number
  sample_count: number
}

/**
 * X-bar / R control chart data with control limits.
 */
export interface SPCChart {
  product_type: string
  measurement_name: string
  center_line: number | null
  ucl: number | null
  lcl: number | null
  r_center: number | null
  r_ucl: number | null
  r_lcl: number | null
  subgroup_size: number
  subgroups: SPCSubgroupStat[]
}

/**
 * An SPC alert raised by anomaly detection.
 */
export interface SPCAlert {
  product_type: string
  measurement_name: string
  rule: string
  severity: string
  message: string
  value: number | null
  timestamp: string
  sample_count: number
}

/**
 * GET /api/v1/spc/{product_type}/{measurement_name} — SPC statistics.
 */
export async function getSPCStatistics(
  productType: string,
  measurementName: string,
  limit = 100,
): Promise<SPCStatistics> {
  const response = await api.get<SPCStatistics>(
    `/spc/${encodeURIComponent(productType)}/${encodeURIComponent(measurementName)}`,
    { params: { limit } },
  )
  return response.data
}

/**
 * GET /api/v1/spc/{product_type}/{measurement_name}/chart — X-bar / R chart data.
 */
export async function getSPCChart(
  productType: string,
  measurementName: string,
  limit = 100,
): Promise<SPCChart> {
  const response = await api.get<SPCChart>(
    `/spc/${encodeURIComponent(productType)}/${encodeURIComponent(measurementName)}/chart`,
    { params: { limit } },
  )
  return response.data
}

/**
 * GET /api/v1/spc/alerts — recent SPC alerts.
 */
export async function getSPCAlerts(limit = 50): Promise<SPCAlert[]> {
  const response = await api.get<SPCAlert[]>('/spc/alerts', { params: { limit } })
  return response.data
}
